internal/services/trackericons/service.go

```diff
@@ -0,0 +1,790 @@
+type Service struct {
+	iconDir string
+	client  *http.Client
+	ua      string
+
+	group singleflight.Group
+
+	failureMu   sync.Mutex
+	lastFailure map[string]time.Time
+}
+
+// Flow overview:
+//   * tracker syncs call QueueFetch when they encounter a host; the helper skips
+//     work if a PNG already exists on disk or the host is still in cooldown.
+//   * GetIcon handles the actual fetch pipeline with singleflight so only one
+//     goroutine ever attempts a download for a given host at a time.
+//   * fetchAndStoreIcon probes a bounded set of candidate URLs, deduplicates
+//     them, attempts each icon once, and stops after the first success. Images
+//     are decoded (falling back to ICO when needed), resized, and written
+//     atomically.
+//   * ListIcons simply streams whatever PNGs already live on disk; the frontend
+//     polls that endpoint but never triggers new fetch attempts.
+//   * Example: encountering "cdn.trackers.example.org" produces host
+//     candidates `["cdn.trackers.example.org", "trackers.example.org",
+//     "example.org"]`. For each we probe `https://…/` and `http://…/`, scrape
+//     any `<link rel="icon">` references, append `/favicon.ico`, and stop at
+//     the first icon that decodes successfully.
+
+// NewService creates a new tracker icon service rooted in the provided data directory.
+func NewService(dataDir, userAgent string) (*Service, error) {
+	if strings.TrimSpace(dataDir) == "" {
+		return nil, fmt.Errorf("data directory must be provided")
+	}
+
+	iconDir := filepath.Join(dataDir, iconDirName)
+	if err := os.MkdirAll(iconDir, 0o755); err != nil {
+		return nil, fmt.Errorf("create tracker icon directory: %w", err)
+	}
+
+	svc := &Service{
+		iconDir:     iconDir,
+		client:      &http.Client{Timeout: fetchTimeout},
+		lastFailure: make(map[string]time.Time),
+	}
+
+	if trimmed := strings.TrimSpace(userAgent); trimmed != "" {
+		svc.ua = trimmed
+	} else {
+		svc.ua = "qui/dev"
+	}
+
+	return svc, nil
+}
+
+func SetGlobal(svc *Service) {
+	globalMu.Lock()
+	globalService = svc
+	globalMu.Unlock()
+}
+
+func QueueFetch(host, trackerURL string) {
+	globalMu.RLock()
+	svc := globalService
+	globalMu.RUnlock()
+
+	if svc == nil {
+		return
+	}
+
+	sanitized := sanitizeHost(host)
+	if sanitized == "" {
+		return
+	}
+
+	// Check if already cached
+	path := svc.iconPath(sanitized)
+	if _, err := os.Stat(path); err == nil {
+		return
+	}
+
+	// Check cooldown
+	if !svc.canAttempt(sanitized) {
+		return
+	}
+
+	// Queue background fetch
+	go func(h string, tracker string) {
+		ctx, cancel := context.WithTimeout(context.Background(), fetchTimeout)
+		defer cancel()
+		if _, err := svc.GetIcon(ctx, h, tracker); err != nil {
+			// Intentionally ignore errors here; they are tracked internally for cooldown.
+		}
+	}(sanitized, trackerURL)
+}
+
+// ListIcons returns all cached tracker icons as base64-encoded data URLs.
+func (s *Service) ListIcons(ctx context.Context) (map[string]string, error) {
```

Comment:
If I could dump a list like this into a txt file (with acronym updated to tracker url), that would be nice.

https://github.com/Audionut/add-trackers/blob/8db05c0e822f9b3afa46ca784644c4e7e400c92b/ptp-add-filter-all-releases-anut.js#L768

Reply:
Most all tracker base64 icons are here
https://github.com/Purfview/IMDb-Scout-Mod/blob/15eb4a80ea076142d59632dc9ab740f17e597d72/IMDb_Scout_Mod.user.js#L2221